components/places/src/ffi.rs

```diff
@@ -113,6 +144,151 @@ impl PlacesConnection {
             )
         })
     }
+
+    /// Add an observation to the database.
+    fn apply_observation(&self, visit: VisitObservation) -> Result<()> {
+        match self.with_conn(|conn| history::apply_observation(conn, visit)) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn get_visited_urls_in_range(
+        &self,
+        start: i64,
+        end: i64,
+        include_remote: bool,
+    ) -> Result<Vec<String>> {
+        self.with_conn(|conn| {
+            history::get_visited_urls(
+                conn,
+                // Probably should allow into()...
+                types::Timestamp(start.max(0) as u64),
+                types::Timestamp(end.max(0) as u64),
+                include_remote,
+            )
+        })
+    }
+
+    fn get_visit_infos(
+        &self,
+        start_date: i64,
+        end_date: i64,
+        exclude_types: VisitTransitionSet,
+    ) -> Result<Vec<HistoryVisitInfo>> {
+        self.with_conn(|conn| {
+            history::get_visit_infos(
+                conn,
+                types::Timestamp(start_date.max(0) as u64),
+                types::Timestamp(end_date.max(0) as u64),
+                exclude_types,
+            )
+        })
+    }
+
+    fn get_visit_count(&self, exclude_types: VisitTransitionSet) -> Result<i64> {
+        self.with_conn(|conn| history::get_visit_count(conn, exclude_types))
+    }
+
+    fn get_visit_page(
+        &self,
+        offset: i64,
+        count: i64,
+        exclude_types: VisitTransitionSet,
+    ) -> Result<Vec<HistoryVisitInfo>> {
+        self.with_conn(|conn| history::get_visit_page(conn, offset, count, exclude_types))
+    }
+
+    fn get_visit_page_with_bound(
+        &self,
+        bound: i64,
+        offset: i64,
+        count: i64,
+        exclude_types: VisitTransitionSet,
+    ) -> Result<HistoryVisitInfosWithBound> {
+        self.with_conn(|conn| {
+            history::get_visit_page_with_bound(conn, bound, offset, count, exclude_types)
+        })
+    }
+
+    // This is identical to get_visited in history.rs but takes a list of strings instead of urls
+    // This is necessary b/c we still need to return 'false' for bad URLs which prevents us from
+    // parsing/filtering them before reaching the history layer
+    fn get_visited(&self, urls: Vec<String>) -> Result<Vec<bool>> {
+        let iter = urls.into_iter();
+        let mut result = vec![false; iter.len()];
+        let url_idxs = iter
+            .enumerate()
+            .filter_map(|(idx, s)| Url::parse(&s).ok().map(|url| (idx, url)))
+            .collect::<Vec<_>>();
+        self.with_conn(|conn| history::get_visited_into(conn, &url_idxs, &mut result))?;
+        Ok(result)
+    }
+
+    fn delete_visits_for(&self, url: Url) -> Result<()> {
+        self.with_conn(|conn| {
+            let guid = match Url::parse(url.as_str()) {
```

Comment:
ditto here re the double parse of the URL, and sadly this probably needs to be a string too, because the failure to parse is consumed.